### PR TITLE
Membership Application: use state machine (aasm gem)

### PR DESCRIPTION
**Pull request by weedySeaDragon**
PT Story:  **Refactor to state machine (big one)**
 https://www.pivotaltracker.com/story/show/134267809

Use the 'aasm' (act as state machine) gem to manage the states, events, and transitions with the MembershipApplication.

### Changes proposed in this pull request:
1.  change 'status' to 'state'
2. use `aasm` gem ('acts as state machine'); put state machine logic into the model (move out of controller)
3. admin uses buttons to change the state of an application (instead of drop-down list). see screenshots (needed to add `.scss` styling for disabled buttons and form (input) buttons)
4. define acceptable states and transitions
5. add more detailed specifications and features relating to the membership application state
6. fixed Swedish wording
7. Proposed a rake task for populating the 'state' column for existing applications
8. Implement the new states and transitions discussed with Sussana, Ulrika, Pat, & Ashley
9. Add a checkbox on the user's edit application page.  If this checkbox is checked when they click 'submit,' the application will change state (from `waiting_for_applicant` to `ready_for_review`)
10. Added some RSpec matchers  to aasm

**State diagram and state transition table for MembershipApplication after applying this PR:**

membership_application

This file is `./docs/membership_application.png`
(This was automatically generated by the aasm_statecharts gem [the version in https://github.com/weedySeaDragon/aasm_statecharts].  Hence the layout isn't the best. )

### Screenshots:

screenshot-aasmstate-1-adminlanding

screenshot-aasmstate-1-admin-membershipstates

screenshot-aasmstate-1-admin-membershipstates-en

screenshot-aasmstate-1-admin-needsinfo-en

#### Checkbox the users sees when she's on the edit form, IF it has been in the `waiting for applicant` state:
(Yes, the label and checkbox need to all be inline.  That can happen in another PR)
screenshot-user-editapp-readyforreviewcheckbox-sv

screenshot-user-editapp-readyforreviewcheckbox

Ready for review:
@thesuss @patmbolger

### CODE REVIEW

File: app/assets/stylesheets/custom-membership-application-state.scss
Comment (thesuss) [reply app/assets/stylesheets/custom-membership-application-state.scss]: ignore this one

File: app/assets/stylesheets/custom-membership-application-state.scss
Comment (thesuss) [reply app/assets/stylesheets/custom-membership-application-state.scss]: ignore this one

**Comment by thesuss** [comment]
Just a note that we have to fix the language before pressing the button on this one.

File: app/models/membership_application.rb
Comment (patmbolger) [inline app/models/membership_application.rb:58]: Please add this line to app/controllers/application_controller.rb:
```
require 'aasm'
```
That will make the module available in rails console.  Without that, I get this error:
```
command: MembershipApplication.all.count
response:  NameError: uninitialized constant MembershipApplication::AASM
```

File: app/models/membership_application.rb
Comment (patmbolger) [inline app/models/membership_application.rb]: Don't need braces if last (or sole) argument to a method is a hash, hence:
```
user.update(is_member: true)
```
reads clearer.

File: app/controllers/membership_applications_controller.rb
Comment (patmbolger) [inline app/controllers/membership_applications_controller.rb]: If the user (applicant) is 

1) editing the application, 
2) and the application is in state `:accepted`, `:rejected`, or `:waiting_for_applicant`
3) the user presses the "Save" button without actually changing anything (did not change any attributes nor upload any files)

then, the event logic:
```
event :applicant_updated_info do
  transitions from: [:accepted, :rejected, :waiting_for_applicant, :pending], to: :pending, guard: :not_a_member?
end
```
will cause a state change to `pending`, even though the user changed nothing.  

In this case, I would think the application should stay in its original state.

If that is the case, this action logic should detect that there are no changes and *not* fire a state change event.

File: app/models/membership_application.rb
Comment (patmbolger) [inline app/models/membership_application.rb:106]: I think that we need to update the `company` for the application *only if* we created a new company(?)

In that case, this statement should be in the ```unless``` block.

File: app/views/membership_applications/_application_status_form.html.haml
Comment (patmbolger) [inline app/views/membership_applications/_application_status_form.html.haml:14]: The event logic for above two buttons `accept` and `ask applicant for info` results in those two buttons being disabled when the admin is viewing a rejected application:
screen shot 2017-01-01 at 9 35 53 am

Yet the state change diagram shows that state change from `rejected` to either of `accepted` or `waiting for applicant` are allowed changes.

File: features/update_membership_application.feature
Comment (patmbolger) [inline features/update_membership_application.feature]: " .... (from waiting for `info` to pending)"

File: features/update_membership_application.feature
Comment (patmbolger) [inline features/update_membership_application.feature]: Might want to change ```needs info``` to ```waiting for applicant```, since that is the actual state name (makes it easier to follow the code and map to state changes).

File: features/update_membership_application.feature
Comment (patmbolger) [inline features/update_membership_application.feature]: Should say:

Admin *can* change from rejected to accepted

File: features/update_membership_application.feature
Comment (patmbolger) [inline features/update_membership_application.feature:290]: I tried to map state changes as shown in the STD to this feature.  In this snapshot, yellow highlights the state changes I did not see in the feature.  (those may be present in other feature tests - I did not look - however, it would be good to have all application state changes tested in this feature).

screen shot 2017-01-01 at 12 07 14 pm

**Review by patmbolger — COMMENTED** [review]
Great work on the PR .... I did not check everything as the PR is quite large - will check more after next iteration.

Please see comments.

File: app/controllers/membership_applications_controller.rb
Comment (thesuss) [reply app/controllers/membership_applications_controller.rb]: Well, wether or not a user CAN/SHOULD edit their application is up for debate. (trying to set up a meeting tomorrow)

File: app/views/membership_applications/_application_status_form.html.haml
Comment (weedySeaDragon) [reply app/views/membership_applications/_application_status_form.html.haml:14]: We haven't decided on the states and transitions yet.  I started with these, but need to change them according to decisions/feedback.

The state can change from `rejected` but only if `not_a_member?` is true.  (In the diagram, the guard conditions for a transition are in square brackets `[]`.)

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb]: agree with both: shouldn't change the state if no changes were actually made, and we need to know if a user can edit their application (probably), and if a member can edit their application (don't know).

File: app/models/membership_application.rb
Comment (weedySeaDragon) [reply app/models/membership_application.rb:106]: Actually, once I thought about this more (thanks for making me do that),  that's not quite right. 

The update line assigns the company instance (whether it was created or already existed) to the membership application.  Until an application is accepted, no company is assigned to that attribute. So that line needs to stay as is.

However, I did improve this by using `find_or_create_by!(company_number: company_number)` instead of the `unless` block.

File: app/controllers/membership_applications_controller.rb
Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb]: Per the meeting we had earlier today (2 January) with Ulrika, we decided that a user must explicitly mark the application as ready for review again; that will change the state to 'Ready for review' (the state formerly known as 'pending').  Just `saving` or even updating information by the user is not enough to change the status.

File: app/models/membership_application.rb
Comment (weedySeaDragon) [reply app/models/membership_application.rb:58]: @patmbolger  I am able to get a correct response without adding that line.  Is it possible you need to run `bundle install` ?  And if you did run bundle install and are still getting the error, then perhaps Susanna can check to see what she gets.

File: app/controllers/membership_applications_controller.rb
Comment (thesuss) [reply app/controllers/membership_applications_controller.rb]: We also have to test that if a user edits their application and adds a category, submits it for review, the new category is not public until approved... 
I don't know if our system caters for this?

Comment (weedySeaDragon) [reply app/controllers/membership_applications_controller.rb]: We don't do this yet.  Sounds like a user story & feature that needs to be written up. :-)

File: app/models/membership_application.rb
Comment (patmbolger) [reply app/models/membership_application.rb:58]: @weedySeaDragon - you're right .... I must have forgotten to run ```bundle```

File: app/models/membership_application.rb
Comment (patmbolger) [reply app/models/membership_application.rb:106]: I did not know that (company not assigned to app until app is accepted).

What is the business rationale for that?

**Review by patmbolger — COMMENTED** [review]
Thanks for comments so far ....

**Comment by thesuss** [comment]
Probably so not to show the potentially added categories on the company until they are indeed added?

**Comment by weedySeaDragon** [comment]
rationale for not having a fully instantiated company until accepted:

Part of the review process is to verify the  Organization number (company number).  (Done mostly by the gem `orgnummer` but also by the SHF admin, I think?)

We only need and use the company information once the applicant has been approved:  visitors can only see companies on the site if they are approved/accepted companies.

If the applicant is not approved/accepted, then there seems no point in instantiating a company and then, when they're rejected, deleting it.  

Make sense, @patmbolger  and @thesuss ?

File: app/models/membership_application.rb
Comment (weedySeaDragon) [reply app/models/membership_application.rb:58]: whew.  glad it was that simple.  Otherwise I was afraid there was something nasty we'd have to untangle. :-)

File: app/models/membership_application.rb
Comment (patmbolger) [reply app/models/membership_application.rb:106]: @weedySeaDragon - I suggest adding a comment in this method to reflect this business rule that a company is not associated with a membership application until that application is accepted.  Otherwise, someone else reading the code might stumble over the same misunderstanding that I had.

Comment (weedySeaDragon) [reply app/models/membership_application.rb:106]: I agree.  In the past, Thomas O. didn't approve of me adding comments to code, so I abstained.  (I didn't understand his boundaries for when it was or wasn't ok.)
I think comments in the code about this are good, but I also think this points to a need for documenting this outside of the code -- where is the right place to do this?  Somewhere on the project page on the AV site?  I think we -- the project as a whole -- need to decide on this.

(edited to add):  Specifications (features & rspecs) are a big part of this; we haven't been using them this way, though.  Should we?

**Comment by weedySeaDragon** [comment]
I know that coverage has declined -- it's because I don't have tests that capture the exceptions raised.  I wanted to go ahead and try to get this wrapped up.  We'll be revising and adding to the states soon and can make sure the coverage increases again.

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Det uppstod ett fel vid godkännande av medlemskap.

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Det uppstod ett fel vid avslag av medlemskap.

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Medlemsansökan har avslagits.

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Ansökan har markerats som: "Behöver kompletteras". (Skicka e-post till den sökande.)

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Det uppstod ett fel när systemet försökte sätta status: "Behöver kompletteras".

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Det uppstod ett fel när systemet försökte ångra status: "Behöver kompletteras".

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Ansökan har ändrats och behöver ej längre kompletteras av sökande.

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Godkänn

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Avböj

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Behöver kompletteras

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Ångra Behöver kompletteras

File: config/locales/sv.yml
Comment (thesuss) [inline config/locales/sv.yml]: Sökanden har kompletterat sin ansökan

File: config/locales/en.yml
Comment (thesuss) [inline config/locales/en.yml]: Used anywhere?

**Review by thesuss — CHANGES_REQUESTED** [review]

**Comment by weedySeaDragon** [comment]
This PR is almost finished (whew).  The 1 remaining thing to be done is to initialize the 'state' column  for existing records. 

I've written a rake task that does this, but haven't committed it yet.   Would this be the best way to handle it, since it's a 1 time thing?  

(I'll commit the rake task in a bit so that it can be reviewed.)

File: config/locales/en.yml
Comment (weedySeaDragon) [reply config/locales/en.yml]: Nope, not anymore.  Good catch.

**Comment by patmbolger** [comment]
Yesterday we had talked about how the application transitions from "waiting for applicant" to "ready for another review" state, once the applicant has finished updating their info.  (that is, we don't transition immediately back to "under review")

However, the state transition table shows a transition to "under review" state:

screen shot 2017-01-12 at 3 42 15 pm

Did I get that wrong?  If not, perhaps we need to remove this particular transition from the PR, or add the state "ready for another review" to accommodate this transition.

**Comment by weedySeaDragon** [comment]
@patmbolger You're right:  --  _where we want to end up_ is with a `ready for review` state.
But this PR isn't there.  We just have the `under_review` state.  (which used to be called "pending")

What this really signals to me is that this PR should either not change any of the states that we had originally, or  else go ahead and implement the new states, i.e. _where we want to end up_ including the `new`, `ready_for_review` states, etc.

My inclination is to go ahead and implement the new states.  Going backwards to where the states were before this PR started would be just as "big" at this point.

Your thoughts?

**Comment by patmbolger** [comment]
I agree .... might as well implement the missing two states as well since you're mostly there anyway.

File: spec/support/ae_aasm_matchers/have_valid_state.rb
Comment (weedySeaDragon) [reply spec/support/ae_aasm_matchers/have_valid_state.rb:29]: do need to have it there so that users can pass in that parameter.

File: spec/support/ae_aasm_matchers/have_valid_transition_from_to.rb
Comment (weedySeaDragon) [reply spec/support/ae_aasm_matchers/have_valid_transition_from_to.rb:24]: do need to have it there so that users can pass in that parameter.

File: spec/support/ae_aasm_matchers/have_valid_transition_from_to.rb
Comment (weedySeaDragon) [reply spec/support/ae_aasm_matchers/have_valid_transition_from_to.rb:28]: do need to have it there so that users can pass in that parameter.

File: spec/support/ae_aasm_matchers/have_valid_state.rb
Comment (weedySeaDragon) [reply spec/support/ae_aasm_matchers/have_valid_state.rb:25]: do need to have it there so that users can pass in that parameter.

File: spec/support/ae_aasm_matchers/have_valid_event.rb
Comment (weedySeaDragon) [reply spec/support/ae_aasm_matchers/have_valid_event.rb:30]: do need to have it there so that users can pass in that parameter.

File: spec/support/ae_aasm_matchers/have_valid_event.rb
Comment (weedySeaDragon) [reply spec/support/ae_aasm_matchers/have_valid_event.rb:26]: do need to have it there so that users can pass in that parameter.

**Review by thesuss — APPROVED** [review]

**Pull request merged**
